### ISSUE TITLE
Regarding how the step method of the learning rate scheduler is used

### ISSUE THREAD

**Issue by nasimrahaman** [facebookresearch/deit#145]
The following assumes that the learning rate schedule is cosine with warmup (i.e. the default). 

`lr_scheduler.step(epoch)` is called [here](https://github.com/facebookresearch/deit/blob/c890ceb9303468cf47553da8764d7febabd9df68/main.py#L382), which comes _after_ `train_one_epoch`. This might lead to unintended behavior, because the first time `train_one_epoch` is called (assuming that the run is not resuming), the learning rate is already at its initial value (which is 1e-6 due to warmup). After that first call to `train_one_epoch`, we still have that `epoch = 0`, meaning `lr_scheduler.step(0)` is essentially what's called. This implies that in the second epoch, the optimizer is still configured to train with the initial learning rate (which is still 1e-6). Only after the second call to `train_one_epoch` is the learning rate bumped up, which means that the schedule is effectively postponed by an epoch (probably not a big deal).  For reference, timm's training script calls [`lr_scheduler.step(epoch + 1)`](https://github.com/rwightman/pytorch-image-models/blob/master/train.py#L643) instead, which sidesteps this issue. 

A related but potentially more important consideration is what happens when a training run is resumed (not for finetuning). If `lr_scheduler.step(epoch)` is called after `train_one_epoch`, `lr_scheduler.step(start_epoch)` must be called before the training loop (e.g. [here](https://github.com/facebookresearch/deit/blob/c890ceb9303468cf47553da8764d7febabd9df68/main.py#L362)). If this is not done, runs that are resuming will always spend an epoch at lr = 1e-6, which might interact in non-trivial ways with the optimizer state (i.e. moving averages). For reference, [here's](https://github.com/rwightman/pytorch-image-models/blob/master/train.py#L488) the relevant line in timm.

**Comment by TouvronHugo** [comment]
Hi @nasimrahaman ,
Thanks for your remark.
For the first point since our results are obtained with this implementation for reasons of reproducibility I will leave the code as is.
For the second point I'll do the modification.
Best,
Hugo
